lib/spack/spack/directives.py

```diff
@@ -932,6 +982,244 @@ def _execute_requires(pkg: Type[spack.package_base.PackageBase]):
     return _execute_requires
 
 
+def remove_all_directive(directive_name):
+    def _remove_all_directive(pkg):
+        getattr(pkg, directive_name).clear()
+
+    return _remove_all_directive
+
+
+@directive("conflicts")
+def drop_all_conflicts():
+    """Removes all conflicts from a package.
+
+    This is typically used when inheriting from another package if the author
+    desires different conflicts than those available in the parent package.
+    """
+    return remove_all_directive("conflicts")
+
+
+@directive("dependencies")
+def drop_all_depends_on():
+    """Removes all depends_on from a package.
+
+    This is typically used when inheriting from another package if the author
+    desires different dependencies than those available in the parent package.
+    """
+    return remove_all_directive("dependencies")
+
+
+@directive("requirements")
+def drop_all_requires():
+    """Removes all requirements from a package.
+
+    This is typically used when inheriting from another package if the author
+    desires different dependencies than those available in the parent package.
+    """
+    return remove_all_directive("requirements")
+
+
+@directive("versions")
+def drop_all_versions():
+    """Removes all versions from a package.
+
+    This is typically used when inheriting from another package if the author
+    desires different versions than those available in the parent package.
+    """
+    return remove_all_directive("versions")
+
+
+class DropDirectiveBase(ABC):
+    name = ""
+
+    def __init__(self, when):
+        self.removal_when = _make_when_spec(when)
+
+    @abstractmethod
+    def add_to_filtered(self, filtered, when, directive_entry):
+        pass
+
+    @abstractmethod
+    def get_directive(self, directive_entry):
+        pass
+
+    def remove(self):
+        removal_when = self.removal_when
+        if not removal_when:
+            return
+        complement_versions = removal_when.versions.complement()
+
+        def _remove(pkg):
+            removal_directive = self.make_directive(pkg)
+            directive_dict = getattr(pkg, self.name)
+            filtered = self.filter_directives(
+                complement_versions, directive_dict, removal_directive, removal_when
+            )
+            directive_dict.clear()
+            directive_dict.update(filtered)
+
+        return _remove
+
+    def filter_directives(
+        self, complement_versions, directive_dict, removal_directive, removal_when
+    ):
+        filtered = {}
+        for when, directives in directive_dict.items():
+            for directive_entry in self.iterate_directives(directives):
+                if self.get_directive(directive_entry) == removal_directive:
```

Comment:
I think this would be clearer if we combine the `get_directive` and `make_directive`  methods into a single method `entry_matches_removal`. Then this line becomes

```suggestion
                if self.entry_matches_removal(directive_entry):
```
and the default implementation for that method could be `return self.get_directive(directive_entry) == self.make_directive(pkg)`, but it gives us more room to rename those methods to something easier to follow as well.